pkg/bff/api/v1/client.go

```diff
@@ -0,0 +1,160 @@
+	var req *http.Request
+	if req, err = s.NewRequest(ctx, http.MethodGet, "/v1/status", nil, nil); err != nil {
+		return nil, err
+	}
+
+	// Execute the request and get a response
+	// NOTE: cannot use s.Do because we want to parse 503 Unavailable errors
+	var rep *http.Response
+	if rep, err = s.client.Do(req); err != nil {
+		return nil, fmt.Errorf("could not execute request: %s", err)
+	}
+	defer rep.Body.Close()
+
+	// Detect other errors
+	if rep.StatusCode != http.StatusOK && rep.StatusCode != http.StatusServiceUnavailable {
+		return nil, fmt.Errorf("[%d] %s", rep.StatusCode, rep.Status)
+	}
+
+	// Deserialize the JSON data from the response
+	out = &StatusReply{}
+	if err = json.NewDecoder(rep.Body).Decode(out); err != nil {
+		return nil, fmt.Errorf("could not deserialize StatusReply: %s", err)
+	}
+	return out, nil
+}
+
+//===========================================================================
+// Helper Methods
+//===========================================================================
+
+const (
+	userAgent    = "GDS BFF API Client/v1"
+	accept       = "application/json"
+	acceptLang   = "en-US,en"
+	acceptEncode = "gzip, deflate, br"
+	contentType  = "application/json; charset=utf-8"
+)
+
+// NewRequest creates an http.Request with the specified context and method, resolving
+// the path to the root endpoint of the API (e.g. /v2) and serializes the data to JSON.
+// This method also sets the default headers of all GDS Admin API v2 client requests.
+func (s *APIv1) NewRequest(ctx context.Context, method, path string, data interface{}, params *url.Values) (req *http.Request, err error) {
+	// Resolve the URL reference from the path
+	endpoint := s.endpoint.ResolveReference(&url.URL{Path: path})
+	if params != nil && len(*params) > 0 {
+		endpoint.RawQuery = params.Encode()
+	}
+
+	var body io.ReadWriter
+	if data != nil {
+		body = &bytes.Buffer{}
+		if err = json.NewEncoder(body).Encode(data); err != nil {
+			return nil, fmt.Errorf("could not serialize request data: %s", err)
+		}
+	} else {
+		body = nil
+	}
+
+	// Create the http request
+	if req, err = http.NewRequestWithContext(ctx, method, endpoint.String(), body); err != nil {
+		return nil, fmt.Errorf("could not create request: %s", err)
+	}
+
+	// Set the headers on the request
+	req.Header.Add("User-Agent", userAgent)
+	req.Header.Add("Accept", accept)
+	req.Header.Add("Accept-Language", acceptLang)
+	req.Header.Add("Accept-Encoding", acceptEncode)
+	req.Header.Add("Content-Type", contentType)
+	return req, nil
+}
+
+// Do executes an http request against the server, performs error checking, and
+// deserializes the response data into the specified struct if requested.
+func (s *APIv1) Do(req *http.Request, data interface{}, checkStatus bool) (rep *http.Response, err error) {
+	if rep, err = s.client.Do(req); err != nil {
+		return rep, fmt.Errorf("could not execute request: %s", err)
+	}
+	defer rep.Body.Close()
+
+	// Detect errors if they've occurred
```

Comment:
Would it be a good idea for us to strongly encourage BFF users to set `checkStatus` to true for debugging/introspection purposes?

Reply:
I'm hoping it will be rare for users of this library to directly call `Do` - instead the library methods that map to the endpoints should be our focus. These methods will generally set `checkStatus` to true, however there are cases where special handling is required (e.g. for a 404 Not Found error). In particular, the `Status` client method will set `checkStatus` to false so that it can handle a 503 Unavailable error, which will still have a body returned from the server.